Mobile: js-draw-based image editor

personalizedrefrigerator opened the pull request:
# Summary
- Adds an editor that allows creating and editing SVG images.
- [js-draw online demo](https://js-draw.web.app/)
- [See the relevant post on the forum](https://discourse.joplinapp.org/t/plugin-js-draw-integration/27114/35?u=personalizedrefriger)

<img alt="screenshot of the image editor in dark mode" src="https://github.com/laurent22/joplin/assets/46334387/e02c16f0-6f50-4578-9402-2e78171a7e3c" height="500"/>

<img alt="screenshot of the image editor in light mode" src="https://github.com/laurent22/joplin/assets/46334387/8fb02aa0-6a45-4a7a-85c6-06da70815b28" height="500"/>

<img alt="screenshot of the image editor in light mode and landscape mode with a menu open" src="https://github.com/laurent22/joplin/assets/46334387/c44e63c4-2442-4108-bb79-adbb9ba92326" width="500"/>


<img alt="screenshot of the image editor in light mode and landscape mode with no menu open" src="https://github.com/laurent22/joplin/assets/46334387/7a42219a-ce80-4360-9891-d5dc6e4d329a" width="500"/>




<img src="https://github.com/laurent22/joplin/assets/46334387/4ebfd5cb-124a-4643-a191-e632211a414e" alt="demonstration of the edit button" width="250"/>

[Screencast from 2023-08-24 06-54-43.webm](https://github.com/laurent22/joplin/assets/46334387/3af898fb-4a12-4bb0-b4a6-98433710bd54)

https://github.com/laurent22/joplin/assets/46334387/b98219fd-15a9-4edb-aac0-3d416404b4e4




# Details
- The image editor can be opened by clicking "Attach" then "Draw Picture" or long-pressing an SVG image and clicking "Open" or clicking the pop-up edit button:
    - <img alt="opening the editor using the properties menu" src="https://user-images.githubusercontent.com/46334387/211168283-dd09ecd5-746a-4c96-97cc-1c24321b5f31.png" height="500"/>
    - <img alt="opening the editor using the edit button" src="https://github.com/laurent22/joplin/assets/46334387/55624e31-a2b2-44a1-8edf-d7a409ec5458" width="250"/>

 
- The editor auto-saves every two minutes.

# Known issues
- In landscape mode, horizontal panning sometimes opens the note properties sidebar instead of moving the js-draw view.
- The `js-draw` selection tool is slow when working with large images.
- In images with a large number of strokes (i.e. when saved, the image is 10+ MiB), `js-draw` can use a very large amount of memory.
- GZipped SVG images aren't supported.


<!--

Please prefix the title with the platform you are targetting:

Here are some examples of good titles:

- Desktop: Resolves #123: Added new setting to change font
- Mobile, Desktop: Fixes #456: Fixed config screen error
- All: Resolves #777: Made synchronisation faster

And here's an explanation of the title format:

- "Desktop" for the Windows/macOS/Linux app (Electron app)
- "Mobile" for the mobile app (or "Android" / "iOS" if the pull request only applies to one of the mobile platforms)
- "CLI" for the CLI app

If it's two platforms, separate them with commas - "Desktop, Mobile" or if it's for all platforms, prefix with "All".

If it's not related to any platform (such as a translation, change to the documentation, etc.), simply don't add a platform.

Then please append the issue that you've addressed or fixed. Use "Resolves #123" for new features or improvements and "Fixes #123" for bug fixes.

AND PLEASE READ THE GUIDE: https://github.com/laurent22/joplin/blob/dev/CONTRIBUTING.md

-->

laurent22 commented:
Thanks for creating the PR! There's a number of error in CI but it looks like it's due to a minified file that's being processed by eslint, so you'd need to add it to eslintignore

laurent22 commented:
I've just tried it again it's quite good in terms of usability, the buttons respond well, and even the sub-options for each buttons are easily accessible.

The issues I've noticed are more to do with design or UX, I list them here and let's see what we can do about it:

- Icons for Pen 1, 2, 3 and Eraser could be improved, as it's not so clear what they represent. Have you checked if there's any open source icon set we could use? If not perhaps we can get a designer to make them
- Spacing and padding between buttons and UI elements (in dialogs) could be improved
- Close and Done button should be separated from the other buttons
- Probably could use something other than black and white for buttons and button states
- Perhaps do a single tap to open the drawing from the note? (or double tap)
- It's such a useful feature that perhaps we could make it more prominent in the context menu (as its own item instead of under "Attach...")?
- Make it more obvious what is the canvas area - maybe like in Photoshop - a white rectangle on a dark gray background?

laurent22 commented on packages/app-mobile/components/NoteEditor/ImageEditor/ImageEditor.tsx:
```diff
@@ -0,0 +1,205 @@
+				svgEditorBundle.listenToolbarState(editor, toolbar);
+
+				// Auto-save every two minutes.
+				const autoSaveInterval = 2 * 60 * 1000;
```
Is it necessary to auto-save? Especially every 2 minutes doesn't seem often enough to be useful? Or maybe it should be more frequent (or not done?)

personalizedrefrigerator replied:
I think auto-saving every two minutes is reasonable. In large images, saving be slow-ish (e.g. 500 ms), so we don't want to do it too often.

I personally have been using `js-draw` to take two-column notes for some of my classes. Because these notes are all in one image, having auto-save is useful.
<details><summary>Example</summary>

![e893808e91a64c5d99b81cbe2f10ad80](https://user-images.githubusercontent.com/46334387/211581109-f917b35a-a5f0-4d27-bae7-acb3b77eb7a2.svg)

</details>

I could, of course, work on making saving faster, cumulative, or asynchronous.

laurent22 replied:
> I think auto-saving every two minutes is reasonable. In large images, saving be slow-ish (e.g. 500 ms), so we don't want to do it too often.

For this kind of usage 2 minutes indeed makes sense so for now let's keep it. But indeed what if it takes a long time to save, let's say more than 120 seconds - what would happen in this case? Should there be some guards to prevent it?

Maybe for now a simple "isSaving" flag and not started to save if it's "true" would be enough.

laurent22 commented on packages/app-mobile/components/NoteEditor/ImageEditor/ImageEditor.tsx:
```diff
@@ -0,0 +1,205 @@
+	const onMessage = useCallback(async (event: WebViewMessageEvent) => {
+		const data = event.nativeEvent.data;
+		if (data.startsWith('error:')) {
+			console.error('ImageEditor:', data);
```
Could you use logger.error() for log messages please? You can create a logger by adding `const logger = Logger.create('ImageEditor')` at the top of the file

laurent22 commented on packages/app-mobile/components/NoteEditor/ImageEditor/ImageEditor.tsx:
```diff
@@ -0,0 +1,205 @@
+		} else if (json.action === 'close') {
+			onRequestCloseEditor();
+		} else {
+			console.error('Unknown action,', json.action);
```
logger.error

laurent22 commented on packages/app-mobile/components/NoteEditor/ImageEditor/ImageEditor.tsx:
```diff
@@ -0,0 +1,205 @@
+	}, [props.onSave, onRequestCloseEditor]);
+
+	const onError = useCallback((event: any) => {
+		console.error('ImageEditor: WebView error: ', event);
```
logger.error

laurent22 commented on packages/app-mobile/components/NoteEditor/ImageEditor/autosave.ts:
```diff
@@ -0,0 +1,24 @@
+import { reg } from '@joplin/lib/registry';
+import shim from '@joplin/lib/shim';
+
+export const autosaveFilename = 'autosaved-drawing.joplin.svg';
```
Since all drawings have the same filename, could there be issues with one drawing loading another drawing file?

personalizedrefrigerator replied:
This is the autosave filename. Only one autosave file is kept (the last one), so if the user accidentally closes the app while the image editor is open (or the OS closes the app to free memory), the last drawing isn't lost.

laurent22 commented on packages/app-mobile/components/NoteEditor/ImageEditor/promptRestoreAutosave.ts:
```diff
@@ -0,0 +1,37 @@
+	if (await shim.fsDriver().exists(autosavePath)) {
+		const title: string|null = null;
+		const message = _(
+			'An autosaved drawing was found. It can either be discarded or attached to the current note.'
```
Maybe just "Joplin has recovered an unsaved drawing"? Then a "Restore it" / "Discard" button below. But I wonder if this message is a bit confusing because what is being restored? What if there are multiple drawings in the note and we overwrite a version of one drawing that the user wanted to keep?

personalizedrefrigerator replied:
The "restore from autosave" functionality should never overwrite an existing drawing. It should always create a copy of the autosaved drawing and attach that to the end of the note (e.g. using `this.attachFile`).

Perhaps "An autosaved drawing was found. Attach a copy of it to the note?".

laurent22 commented on packages/app-mobile/components/screens/Note.tsx:
```diff
@@ -748,6 +764,56 @@ class NoteScreenComponent extends BaseScreenComponent {
+
+		const filePath = resourcePath ?? `${Setting.value('resourceDir')}/saved-drawing.joplin.svg`;
+		await shim.fsDriver().writeFile(filePath, svgData, 'utf8');
+		console.info('Saved drawing to', filePath);
```
logger.info

personalizedrefrigerator commented:
> Icons for Pen 1, 2, 3 and Eraser could be improved, as it's not so clear what they represent. Have you checked if there's any open source icon set we could use? If not perhaps we can get a designer to make them

At present, the pen icons resize/change color to match the pen thickness/color. While we could reasonably use font-awesome or similar icons for these, we wouldn't have this size/color change.
![image](https://user-images.githubusercontent.com/46334387/211583032-c5aff148-4e4e-4351-95dd-baceef4c69b7.png)
![image](https://user-images.githubusercontent.com/46334387/211583107-8e4d5032-45e7-40ad-b98f-f518d8e92564.png)
![image](https://user-images.githubusercontent.com/46334387/211583172-826439fe-0eeb-4e6f-bd06-8c3a01eaeb90.png)

I do agree that these icons don't look very similar to pens/erasers (so a re-design makes sense). I'll try re-designing them, then ask for feedback/help.

personalizedrefrigerator commented:
> Probably could use something other than black and white for buttons and button states

The black/white are taken from Joplin's current theme. For example, with the Dracula theme,
![image](https://user-images.githubusercontent.com/46334387/211585040-95f14179-7ff7-46d9-91d2-481000025380.png)

It might make sense to adjust the colors of the content of some of the icons, though. For example, the image icon could use green/yellow for the content of the image.

personalizedrefrigerator commented:
> Perhaps do a single tap to open the drawing from the note? (or double tap)

I worry about the feature being difficult to discover and/or too easy to accidentally trigger. Loading can be slow with large images, so I'm against a single-tap. Perhaps we could do something similar to the desktop app and show an "edit" button in the corner of the image when the user taps once on the image.

laurent22 commented:
> Perhaps we could do something similar to the desktop app and show an "edit" button in the corner of the image when the user taps once on the image.

Do you know what other mobile apps handle this in general? Only one I could try here is Samsung Notes but they don't have this concept of separate drawing area. The "Edit" button indeed seems the most intuitive way and shouldn't result in accidental taps.

laurent22 commented:
> At present, the pen icons resize/change color to match the pen thickness/color.

I hadn't realised that, but now indeed I see it. I'm not certain we need to display the pen size within the icon because even then it's not really obvious what the real size will be. For example I was confused when seeing the arrow at a certain size within the icon, but at a completely different size once it's drawn (and of course the real size wouldn't fit within the icon).

Since there are many drawing apps out there maybe we could check what they do? In the few that's I've checked, they usually simply display the type of brush in the icon and nothing more, which seems reasonable to me.

personalizedrefrigerator commented:
> Do you know what other mobile apps handle this in general? Only one I could try here is Samsung Notes but they don't have this concept of separate drawing area. The "Edit" button indeed seems the most intuitive way and shouldn't result in accidental taps.

Google Docs allows embedding drawings and shows an "edit" button in the menu that appears when the drawing is selected:
![](https://user-images.githubusercontent.com/46334387/211972767-e72995cf-f7d2-499f-8dbf-52978da48dce.png)

Something similar to this could make sense.

laurent22 commented:
> Something similar to this could make sense.

Yes I think just that edit button would be very good. But we can also keep it in the context menu since that's consistent with how images are shared for example

personalizedrefrigerator commented:
Icons are in the process of being re-designed: https://github.com/personalizedrefrigerator/js-draw/pull/14

laurent22 commented:
Those new icons look great!

personalizedrefrigerator commented on packages/app-mobile/components/NoteBodyViewer/hooks/useEditPopup.test.ts:
```diff
@@ -0,0 +1,57 @@
+};
+
+const destroyEditPopup = () => {
+	const { destroyEditPopupSyntax } = getEditPopupSource();
```
This needs to be refactored — I don't like exporting both `getEditPopupSource` and `useEditPopup` from `useEditPopup.ts`. Perhaps the code for the edit popup can live in `packages/renderer`?

personalizedrefrigerator replied:
At least for now, `useEditPopup` makes sense in `packages/app-mobile`:
1. `packages/renderer` currently doesn't seem to be linked to `@joplin/lib`, thus localizing strings with `_(...)` will be difficult.
2. `useEditPopup` returns CSS in addition to JavaScript strings. With `useEditPopup` structured as it is currently, it's easy to inject the CSS into the WebView. Moving calls to `useEditPopup` to the renderer might make this more difficult.

personalizedrefrigerator commented on packages/app-mobile/components/NoteBodyViewer/hooks/useOnMessage.ts:
```diff
@@ -1,7 +1,25 @@
+) {
+	const {
+		onMarkForDownload, onResourceLongPress, onCheckboxChange, onRequestEditResource, onJoplinLinkClick,
+	} = callbacks;
```
`callbacks` is destructured into component variables here to prevent the dependency hook from requesting that the `useCallback` depend on `callbacks` (which, because `({ a: 1 }) !== ({ a: 1 })`, would cause the `useCallback`'s dependencies to change every time `useOnMessage` is called).

To prevent this change from being undone, I should add a comment here.

personalizedrefrigerator commented on packages/app-mobile/components/NoteEditor/ImageEditor/ImageEditor.tsx:
```diff
@@ -0,0 +1,215 @@
+		window.saveDrawing = saveDrawing;
+
+		const closeDrawing = () => {
+			window.ReactNativeWebView.postMessage('{ "action": "close" }');
```
When there are no changes to the drawing (the undo stack is empty), a different action should be fired, one that doesn't show an "are you sure you want to close" dialog.

personalizedrefrigerator reviewed (commented):
I've left comments on things I plan to change!

personalizedrefrigerator commented on packages/renderer/MdToHtml/createEventHandlingAttrs.ts:
```diff
@@ -0,0 +1,101 @@
+
+// Adds event-handling (e.g. long press) code to images and links.
+// resourceId is the ID of the image resource or link.
+const createEventHandlingAttrs = (resourceId: string, options: Options, onClickAction: string|null) => {
```
This refactor makes it much easier to add an `edit` popup, but it should be moved to a separate PR. (I plan to do so later this week).

personalizedrefrigerator replied:
See #7774

personalizedrefrigerator commented on packages/lib/components/shared/note-screen-shared.ts:
```diff
@@ -21,7 +21,7 @@ interface Shared {
 	toggleIsTodo_onPress?: (comp: any)=> void;
 	toggleCheckboxRange?: (ipcMessage: string, noteBody: string)=> any;
-	toggleCheckbox?: (ipcMessage: string, noteBody: string)=> void;
+	toggleCheckbox?: (ipcMessage: string, noteBody: string)=> string;
```
This change should be moved to a separate PR. Note that the string return type is used by `useOnMessage.ts`.

I plan to open a pull request for this (and related return-type changes) some time in the next few days.

personalizedrefrigerator commented:
Images now have a default, mostly-transparent background-color. This makes it clearer where the canvas is:

<img src="https://user-images.githubusercontent.com/46334387/220718259-f5b26aea-7f46-49ad-b61e-5d570258bf5c.png" alt="screenshot showing the new background color in the editor, light-mode" width=300/> <img src="https://user-images.githubusercontent.com/46334387/220720253-edc7efea-fa3f-4176-acc3-c87a89edce91.png" alt="screenshot showing the new background color in the editor, dark-mode" width=300/>

If the user changes the background color in "Document" > "Background Color", that background color is used for new images.

personalizedrefrigerator commented:
# Progress summary 

> The issues I've noticed are more to do with design or UX, I list them here and let's see what we can do about it:

Below is a summary of how the above feedback has been addressed/not addressed.

> - Icons for Pen 1, 2, 3 and Eraser could be improved, as it's not so clear what they represent. Have you checked if there's any open source icon set we could use? If not perhaps we can get a designer to make them.
> - Spacing and padding between buttons and UI elements (in dialogs) could be improved

This has been partially addressed. Of course, the icons can still be improved.
![image](https://user-images.githubusercontent.com/46334387/222522340-12ab469c-28cc-4eb5-8dd1-70c7e25715dd.png)


> - Close and Done button should be separated from the other buttons

![image](https://user-images.githubusercontent.com/46334387/222532013-d9e74995-e116-4e8d-a7ae-d149054aab7b.png)

> - Probably could use something other than black and white for buttons and button states

Black and white are taken from Joplin's default theme. More specifically, they have `background: theme.selectionColor2` and `color: theme.color2`. While this results in black-on-white for selection in the default light theme,
![image](https://user-images.githubusercontent.com/46334387/222536620-44a8a8e9-4da1-45f8-977f-9c2549129f7e.png)
it seems to work well in other themes:
![image](https://user-images.githubusercontent.com/46334387/222537823-befc89e3-0f6d-4a87-b843-380a976ac041.png)
![image](https://user-images.githubusercontent.com/46334387/222537158-d48e3784-98aa-459d-bdbe-101be5779d66.png)
![image](https://user-images.githubusercontent.com/46334387/222537269-fd040884-7ebd-41b8-9867-6dc23e6d6fe1.png)
![image](https://user-images.githubusercontent.com/46334387/222537388-52a261f2-be45-47e9-9726-b4fcd136db1a.png)
![image](https://user-images.githubusercontent.com/46334387/222537571-f4e5b475-1e0a-4c1e-9f1a-e9a39288876e.png)
![image](https://user-images.githubusercontent.com/46334387/222537707-257a5c2c-e0d9-44ea-841d-45d4c4c43b38.png)
![OLED Dark: Blue&white selection](https://user-images.githubusercontent.com/46334387/222537949-caf9635f-6c74-4eee-8df2-a9bdeeacb052.png)


> - Perhaps do a single tap to open the drawing from the note? (or double tap)

Images now have a pop-up edit button that appears after tapping on the image once. This is in addition to an "edit" option when long-pressing:
![image](https://user-images.githubusercontent.com/46334387/222538200-506bcc9d-6446-4e8a-b79e-145e46c7c60b.png)


> - It's such a useful feature that perhaps we could make it more prominent in the context menu (as its own item instead of under "Attach...")?

I think I prefer keeping "Insert Drawing" under the attach menu. With "Insert Drawing" directly in the "⠇" menu, the menu starts to look over-full/cluttered:
![image](https://user-images.githubusercontent.com/46334387/222539798-06b4d6d5-6088-4d14-b07d-d962d7e0c97e.png)

Additionally, the "attach" menu is accessible from both the markdown toolbar and the context ("⠇") menu. Thus, it shouldn't be too difficult to insert a drawing. Of course, it might make sense to test this claim.

> - Make it more obvious what is the canvas area - maybe like in Photoshop - a white rectangle on a dark gray background?

I've accomplished this with a mostly-transparent default background. I could also make the default background color white and not transparent. The transparency is to allow drawings to look dark in dark mode and light in light mode (the editor's default colors are chosen to be visible in both dark/light modes).

personalizedrefrigerator commented:
**Progress update**: The toolbar UI is in the process of being re-implemented [based on this design](https://github.com/laurent22/joplin/issues/8568). [The new UI can be previewed here](https://js-draw.web.app/).

Wladefant commented:
Wow nice. Didn't except this feature after five years of no movement. But this is really nice. Joplin would be the only note taking app with this kind of feature. @personalizedrefrigerator when are you planning to release this feature?

personalizedrefrigerator commented:
> Wow nice. Didn't except this feature after five years of no movement. But this is really nice. Joplin would be the only note taking app with this kind of feature. @personalizedrefrigerator when are you planning to release this feature?

At present, `js-draw` is undergoing design review (#8568), so hopefully this will be released in at most a few months!

You can leave feedback based on either the [pre-release](https://github.com/personalizedrefrigerator/joplin/releases) or the [web version of js-draw](https://js-draw.web.app/demo/index.html) here or in the discussions/issues section of the [js-draw GitHub repository](https://github.com/personalizedrefrigerator/js-draw).

personalizedrefrigerator commented:
I've changed how the save/exit buttons work ([related Figma comment](https://www.figma.com/file/NA5F2AMWO3wUuaoDfUaAb8?node-id=55025:27099&mode=design#563673580)):

https://github.com/laurent22/joplin/assets/46334387/85ffb5d2-5d85-42c3-b94e-2e463f5f425e

This pull request should now be ready for review!

laurent22 commented:
- Is it possible remove the scrollbar on the toolbar? (I feel it's not that useful and is drawn over the icons)

<img width="155" alt="image" src="https://github.com/laurent22/joplin/assets/1285584/7ad428a7-23fb-47b6-93db-a0b61ab7199a">

- When clicking on a drawing, it shows this question mark icon:

<img width="116" alt="image" src="https://github.com/laurent22/joplin/assets/1285584/9adef066-d4a6-4a4b-8dd2-727724e260c1">

- That button also appears in seemingly random places. In some cases, the image cannot be edited because the button is completely off-screen.

<img width="214" alt="image" src="https://github.com/laurent22/joplin/assets/1285584/195c382d-e050-4f3b-8020-6deb430cd0a1">

- Would it make sense to call the Save/Exit button Save/Cancel? Also please remove the icons from these buttons

- I've just checked a few iOS and Android apps and the Save button is usually on the right, and Cancel on the left so please swap the order

- When dragging the toolbar, the tooltips should not appear

- The grid shouldn't be displayed in the note viewer

- Could you put the menu item below "Attach...", to make it easier to discover

The rest looks very good, thanks Henry! I think we're close to a release with this.

personalizedrefrigerator commented:
> Is it possible remove the scrollbar on the toolbar? (I feel it's not that useful and is drawn over the icons)

This should be done!

Note that I'm not applying this change upstream [because of possible accessibility concerns](https://developer.mozilla.org/en-US/docs/Web/CSS/scrollbar-width#accessibility_concerns). (It should be fine here, though, because we can assume Joplin is running on a touchscreen device).

> When clicking on a drawing, it shows this question mark icon:

Fixed! This had to do with where the icons were stored and should now be fixed! Previously, icons weren't readable by the WebView on iOS.

> That button also appears in seemingly random places. In some cases, the image cannot be edited because the button is completely off-screen.

This is also fixed! This was caused by `absolute` positioning — it seems that in the iOS WebView, `absolute` positioned buttons don't scroll with the container. 

> Would it make sense to call the Save/Exit button Save/Cancel? Also please remove the icons from these buttons
>
>  I've just checked a few iOS and Android apps and the Save button is usually on the right, and Cancel on the left so please swap the order

As per the [discussion here](https://github.com/personalizedrefrigerator/js-draw/discussions/40), I've renamed "Exit" to "Close", rather than "Cancel".

![screenshot: close on left save on right](https://github.com/laurent22/joplin/assets/46334387/476ddf1a-bfea-4337-b080-ffd9f2027d61)

I've also removed the icons, but only in portrait mode. In landscape mode (on small screens), the editor looks like this:
![screenshot: close and save have no icons](https://github.com/laurent22/joplin/assets/46334387/3c6009d2-ba01-4c88-920a-97230b90b66b)

I've also updated the editor to use localized strings from Joplin for the save and close buttons — at present, Joplin is far better localized than `js-draw`.

Because the save and close icons hint at functionality (which is important when a localization is missing), I've left the icons in the upstream `js-draw` portrait-mode editor, but have made them smaller. 
<img src="https://github.com/laurent22/joplin/assets/46334387/c1ad578b-e16f-4945-8508-5e75af044edd" width="300" alt="screenshot: icons kept"/>

I did, however, decrease the size of the icons so that they are roughly the same size as the text.

> When dragging the toolbar, the tooltips should not appear

I've made this better, but I don't think it's fixed entirely (scrolling very slowly still opens the tooltips). Currently tooltips are shown/hidden with CSS. 

> The grid shouldn't be displayed in the note viewer

For now, I've disabled the grid by default in `js-draw` instead. Disabling the grid by default was [also requested here](https://github.com/personalizedrefrigerator/js-draw/discussions/36) (though this does mean that the drawing canvas will initially appear empty).

Preserving the grid background is similar in functionality to:
- Xournal++

It is different from:
- OneNote (tested with the online version, print to PDF)
- FreeForm (tested on iOS, share as PDF)
- Inkscape

I would personally like to keep the grid on export (it's useful for drawing graphs), but given that OneNote, FreeForm, and Inkscape all act differently, it's probably best to hide the grid.

I expect making this change in `js-draw` to require some refactoring (which I plan to have done in a day or two).

> Could you put the menu item below "Attach...", to make it easier to discover

Done!
![image](https://github.com/laurent22/joplin/assets/46334387/f6fdc348-a0f8-466e-a77f-3a1284fb59bd)

Wladefant commented on packages/app-mobile/components/NoteEditor/ImageEditor/js-draw/applyTemplateToEditor.ts:
```diff
@@ -18,9 +18,7 @@ const applyTemplateToEditor = async (editor: Editor, templateData: string) => {
-					'backgroundType': BackgroundComponentBackgroundType.Grid,
-					'gridSize': 25,
-					'gridStrokeWidth': 0.7,
+					'backgroundType': BackgroundComponentBackgroundType.SolidColor,
```
That is changing the grid to solid in the viewer, am I correct?

personalizedrefrigerator replied:
Yes. See ["The grid shouldn't be displayed in the note viewer" section](https://github.com/laurent22/joplin/pull/7588#issuecomment-1738257209) in the above comment.

laurent22 commented:
Ok that looks good now. I guess there might be a few tweaks to make here and there as the component becomes more widely tested during pre-release, but it's already very good as it is. Thanks a lot for adding this functionality Henry!

rnbergren commented:
been using this for about a week. Testing it works quite well. Slightly clunky getting used to the first use.

Only update I would like to see is the button on my pen being automagically the erase functionality. I don't believe it is mapped to anything yet. I am using a Samsung Fold 5 and Samsung S7 tablet. 

Kudos for doing this. Makes this a complete OneNote killer.

personalizedrefrigerator commented:
> Only update I would like to see is the button on my pen being automagically the erase functionality. I don't believe it is mapped to anything yet. I am using a Samsung Fold 5 and Samsung S7 tablet.

Thank you for the feedback!

I would also like the eraser on my tablet's pen to work, however, the pen doesn't seem to send events to the editor when both the eraser button and pen are down (only when the eraser button is down and the pen is above the screen). You can see the events currently received by js-draw [using this debug tool (run in Chrome or similar for rough equivalence with how it runs within Joplin)](https://js-draw.web.app/debugging/stroke-logging/).

I've considered making the eraser button erase while the pen is above the screen, but that would likely be confusing (as the pen would do nothing when touching the screen).

paoloap commented:
This is absolutely a game changing feature.
Unfortunately my Pixel is broken and I'm stuck with an iPhone. Is there a way to try a version of the app with this function on iOS?

personalizedrefrigerator commented:
> This is absolutely a game changing feature. Unfortunately my Pixel is broken and I'm stuck with an iPhone. Is there a way to try a version of the app with this function on iOS?

Yes! As of today, it's available on TestFlight. See [the forum post](https://discourse.joplinapp.org/t/a-new-ios-beta-version-is-available/32361).

MMachado05 commented:
This is awesome! Is there a similar drawing capability on the desktop app? I've been looking around but can't seem to find info on one.

personalizedrefrigerator commented:
Yes! It's distributed as a plugin: https://joplin.github.io/website-plugin-discovery/plugin/io.github.personalizedrefrigerator.js-draw/

Ben-Kempton commented:
> > Only update I would like to see is the button on my pen being automagically the erase functionality. I don't believe it is mapped to anything yet. I am using a Samsung Fold 5 and Samsung S7 tablet.
> 
> Thank you for the feedback!
> 
> I would also like the eraser on my tablet's pen to work, however, the pen doesn't seem to send events to the editor when both the eraser button and pen are down (only when the eraser button is down and the pen is above the screen). You can see the events currently received by js-draw [using this debug tool (run in Chrome or similar for rough equivalence with how it runs within Joplin)](https://js-draw.web.app/debugging/stroke-logging/).
> 
> I've considered making the eraser button erase while the pen is above the screen, but that would likely be confusing (as the pen would do nothing when touching the screen).

Could it be something like a tool switch when the pen is above the screen and the button is pressed? Ex, writing with one of the pen tools, lift pen from screen, click button, switches to eraser, put pen back on screen to use eraser normally, lift pen again, click button, and toggles back to previous tool?  Been loving the drawing support so far!
